core/src/main/java/kafka/server/SharePartition.java

```diff
@@ -310,37 +485,264 @@ private void initialize() {
+        private final long expirationMs;
+        private final String memberId;
         private final long firstOffset;
-        /**
-         * The last offset of the batch that is fetched from the log.
-         */
         private final long lastOffset;
+
+        AcquisitionLockTimerTask(long delayMs, String memberId, long firstOffset, long lastOffset) {
+            super(delayMs);
+            this.expirationMs = time.hiResClockMs() + delayMs;
+            this.memberId = memberId;
+            this.firstOffset = firstOffset;
+            this.lastOffset = lastOffset;
+        }
+
+        long expirationMs() {
+            return expirationMs;
+        }
+
         /**
-         * The in-flight state of the fetched records. If the offset state map is empty then inflightState
-         * determines the state of the complete batch else individual offset determines the state of
-         * the respective records.
+         * The task is executed when the acquisition lock timeout is reached. The task releases the acquired records.
          */
+        @Override
+        public void run() {
+            releaseAcquisitionLockOnTimeout(memberId, firstOffset, lastOffset);
+        }
+    }
+
+    private void releaseAcquisitionLockOnTimeout(String memberId, long firstOffset, long lastOffset) {
+        // TODO: Implement the logic to release the acquisition lock on timeout.
+    }
+
+    private boolean checkForFullMatch(InFlightBatch inFlightBatch, long firstOffsetToCompare, long lastOffsetToCompare) {
+        return inFlightBatch.firstOffset >= firstOffsetToCompare && inFlightBatch.lastOffset <= lastOffsetToCompare;
+    }
+
+    // Visible for testing. Should only be used for testing purposes.
+    NavigableMap<Long, InFlightBatch> cachedState() {
+        return new ConcurrentSkipListMap<>(cachedState);
+    }
+
+    /**
+     * The InFlightBatch maintains the in-memory state of the fetched records i.e. in-flight records.
+     */
+    final class InFlightBatch {
+        // The offset of the first record in the batch that is fetched from the log.
+        private final long firstOffset;
+        // The last offset of the batch that is fetched from the log.
+        private final long lastOffset;
+
+        // The in-flight state of the fetched records. If the offset state map is empty then inflightState
+        // determines the state of the complete batch else individual offset determines the state of
+        // the respective records.
         private InFlightState inFlightState;
 
-        InFlightBatch(String memberId, long firstOffset, long lastOffset, RecordState state, int deliveryCount) {
+        // The subset of offsets within the batch that holds different record states within the parent
+        // fetched batch. This is used to maintain the state of the records per offset, it is complex
+        // to maintain the subset batch (InFlightBatch) within the parent batch itself as the nesting
```

Comment:
I'm not sure I understand the point about nesting in this comment.

Reply:
Yeah, you are right. This statement doesn't make much sense as I was initially thinking to divide batches and wrote about nesting. I have re-word it.